src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProjectFactory.cs

```diff
@@ -0,0 +1,657 @@
+        {
+            var id = ProjectId.CreateNewId(projectSystemName);
+            var assemblyName = creationInfo.AssemblyName ?? projectSystemName;
+
+            // We will use the project system name as the default display name of the project
+            var project = new ProjectSystemProject(
+                this,
+                hostInfo,
+                id,
+                displayName: projectSystemName,
+                language,
+                assemblyName: assemblyName,
+                compilationOptions: creationInfo.CompilationOptions,
+                filePath: creationInfo.FilePath,
+                parseOptions: creationInfo.ParseOptions);
+
+            var versionStamp = creationInfo.FilePath != null ? VersionStamp.Create(File.GetLastWriteTimeUtc(creationInfo.FilePath))
+                                                             : VersionStamp.Create();
+
+            await ApplyChangeToWorkspaceAsync(w =>
+            {
+                var projectInfo = ProjectInfo.Create(
+                    new ProjectInfo.ProjectAttributes(
+                        id,
+                        versionStamp,
+                        name: projectSystemName,
+                        assemblyName: assemblyName,
+                        language: language,
+                        checksumAlgorithm: SourceHashAlgorithms.Default, // will be updated when command line is set
+                        compilationOutputFilePaths: default, // will be updated when command line is set
+                        filePath: creationInfo.FilePath,
+                        telemetryId: creationInfo.TelemetryId),
+                    compilationOptions: creationInfo.CompilationOptions,
+                    parseOptions: creationInfo.ParseOptions);
+
+                // If we don't have any projects and this is our first project being added, then we'll create a new SolutionId
+                // and count this as the solution being added so that event is raised.
+                if (w.CurrentSolution.ProjectIds.Count == 0)
+                {
+                    w.OnSolutionAdded(
+                        SolutionInfo.Create(
+                            SolutionId.CreateNewId(SolutionPath),
+                            VersionStamp.Create(),
+                            SolutionPath,
+                            projects: new[] { projectInfo },
+                            analyzerReferences: w.CurrentSolution.AnalyzerReferences)
+                        .WithTelemetryId(SolutionTelemetryId));
+                }
+                else
+                {
+                    w.OnProjectAdded(projectInfo);
+                }
+            }).ConfigureAwait(false);
+
+            return project;
+        }
+
+        public string? TryGetDependencyNodeTargetIdentifier(ProjectId projectId)
```

Comment:
The members here to the end of the file are moved out of VIsualStudioWorkspaceImpl; I have kept the ordering of the members the same for ease of review and making it a bit easier on Git to track any merges across it, but we do need to do a follow-up to make the ordering more sane. There are fields mixed into this, for example, which is icky.